pkg/statistics/handle/syncload/stats_syncload.go

```diff
@@ -223,46 +226,59 @@ func (s *statsSyncLoad) HandleOneTask(sctx sessionctx.Context, lastTask *statsty
 	} else {
 		task = lastTask
 	}
+	result := stmtctx.StatsLoadResult{Item: task.Item.TableItemID}
 	resultChan := s.StatsLoad.Singleflight.DoChan(task.Item.Key(), func() (any, error) {
-		return s.handleOneItemTask(sctx, task)
+		err := s.handleOneItemTask(sctx, task)
+		return nil, err
 	})
 	timeout := time.Until(task.ToTimeout)
 	select {
-	case result := <-resultChan:
-		if result.Err == nil {
-			slr := result.Val.(*stmtctx.StatsLoadResult)
-			if slr.Error != nil {
-				return task, slr.Error
-			}
-			task.ResultCh <- *slr
+	case sr := <-resultChan:
+		// sr.Val is always nil.
+		if sr.Err == nil {
+			task.ResultCh <- result
+			return nil, nil
+		}
+		if !updateNeededItemTaskRetryCountAndCheck(task) {
+			result.Error = sr.Err
+			task.ResultCh <- result
 			return nil, nil
 		}
-		return task, result.Err
+		return task, sr.Err
 	case <-time.After(timeout):
+		if !updateNeededItemTaskRetryCountAndCheck(task) {
+			result.Error = errors.New("stats loading timeout")
+			task.ResultCh <- result
+			return nil, nil
+		}
 		task.ToTimeout.Add(time.Duration(sctx.GetSessionVars().StatsLoadSyncWait.Load()) * time.Microsecond)
 		return task, nil
 	}
 }
 
-func (s *statsSyncLoad) handleOneItemTask(sctx sessionctx.Context, task *statstypes.NeededItemTask) (result *stmtctx.StatsLoadResult, err error) {
+func updateNeededItemTaskRetryCountAndCheck(task *statstypes.NeededItemTask) bool {
```

Comment:
```suggestion
func isVaildForRetry(task *statstypes.NeededItemTask) bool {
```

Or any more straight name.

Reply:
Done